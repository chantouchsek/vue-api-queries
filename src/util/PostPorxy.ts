import BaseProxy from '../core/BaseProxy'

class PostProxy extends BaseProxy {
  constructor(parameters = {}) {
    super('posts', parameters)
  }

  tags<T>(id: string | number) {
    return this.submit<T>('get', `${id}/tags`)
  }

  throwException<T>(id: string | number) {
    return this.submit<T>('unlink', `${id}/tags`)
  }
}

export default PostProxy
