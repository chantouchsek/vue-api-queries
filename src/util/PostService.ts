import BaseService from '../core/BaseService'

class PostService extends BaseService {
  constructor(parameters = {}) {
    super('posts', parameters)
  }

  tags<T = any>(id: string | number) {
    return this.submit<T>('get', `${id}/tags`)
  }
}

export default PostService
