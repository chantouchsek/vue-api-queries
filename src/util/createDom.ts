export function disableTransitions() {
  const css = `
  * {
    animation: none !important;
    transition: none !important;
    transition-timing-function: none !important;
  }`
  const head = document.head || document.getElementsByTagName('head')[0]
  const style = document.createElement('style')
  head.appendChild(style)
  style.type = 'text/css'
  style.appendChild(document.createTextNode(css))
}

export function addElemWithDataAppToBody() {
  const app = document.createElement('div')
  app.setAttribute('data-app', String(true))
  document.body.append(app)
}
