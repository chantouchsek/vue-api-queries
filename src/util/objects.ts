export function isArray(object: any): boolean {
  return Object.prototype.toString.call(object) === '[object Array]'
}

export function hasOwnProperty(obj: any, key: any) {
  if (!obj || !key) return false
  return Object.prototype.hasOwnProperty.call(obj, key)
}

export function isFile(object: any): boolean {
  if (
    typeof window === 'undefined' ||
    typeof File !== 'function' ||
    typeof FileList !== 'function'
  ) {
    return false
  }
  return object instanceof File || object instanceof FileList
}

export function merge(a: any, b: any): Record<any, any> {
  for (const key in b) {
    if (hasOwnProperty(b, key)) {
      a[key] = cloneDeep(b[key])
    }
  }
  return a
}

export function cloneDeep(object: any): any {
  if (object === null) {
    return null
  }

  if (isFile(object)) {
    return object
  }

  if (isArray(object)) {
    const clone: any = []

    for (const key in object) {
      if (hasOwnProperty(object, key)) {
        clone[key] = cloneDeep(object[key])
      }
    }

    return clone
  }

  if (typeof object === 'object') {
    const clone: any = {}

    for (const key in object) {
      if (hasOwnProperty(object, key)) {
        clone[key] = cloneDeep(object[key])
      }
    }

    return clone
  }

  return object
}

export function is(errors: any, error: any): boolean {
  return isArray(error)
    ? error.some((w: string) => is(errors, w))
    : errors.includes(error)
}
