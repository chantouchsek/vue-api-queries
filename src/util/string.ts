export function removeDoubleSlash(url: string) {
  return url.replace(/\/\//g, '/')
}
