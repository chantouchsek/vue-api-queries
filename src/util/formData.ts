import { hasOwnProperty, isArray, isFile } from './objects'

const getKey = (parent: any, property: any) => (parent ? parent + '[' + property + ']' : property)
export function objectToFormData(obj: any, formData = new FormData(), parent = ''): FormData {
  if (obj === null || obj === 'undefined' || obj.length === 0) {
    formData.append(parent, obj)
  } else {
    for (const property in obj) {
      if (hasOwnProperty(obj, property)) {
        appendToFormData(formData, getKey(parent, property), obj[property])
      }
    }
  }
  return formData
}
function appendToFormData(formData: FormData, key: string, value: any) {
  if (value instanceof Date) return formData.append(key, value.toISOString())
  if (value instanceof File) return formData.append(key, value, value.name)
  if (typeof value === 'boolean') return formData.append(key, value ? '1' : '0')
  if (value === null) return formData.append(key, '')
  if (typeof value !== 'object') return formData.append(key, value)
  if (isArray(value) && hasFilesDeep(value)) {
    for (let i = 0; i < value.length; i++) {
      formData.append(key + '[' + i + ']', value[i], value[i].name)
    }
    return formData
  }
  objectToFormData(value, formData, key)
}
export function hasFilesDeep(obj: any): boolean {
  if (obj === null) return false
  if (typeof obj === 'object') {
    for (const key in obj) {
      if (isFile(obj[key])) return true
    }
  }
  if (isArray(obj)) {
    let f = ''
    for (const key in obj) {
      if (hasOwnProperty(obj, key)) {
        f = key
        break
      }
    }
    return hasFilesDeep(obj[f])
  }
  return isFile(obj)
}
export function hasFiles(form: any) {
  for (const prop in form) {
    const hasProp = hasOwnProperty(form, prop) || typeof window !== 'undefined'
    if (hasProp && hasFilesDeep(form[prop])) return true
  }
  return false
}
