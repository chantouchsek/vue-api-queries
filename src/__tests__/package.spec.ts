import Vue from 'vue'
import VueApiQueries from '../'
import { addElemWithDataAppToBody, disableTransitions } from '../util/createDom'
import { sleep } from '../util/promise'
import { mount } from '@vue/test-utils'

describe('Vue Api Queries', () => {
  disableTransitions()
  addElemWithDataAppToBody()
  Vue.use(VueApiQueries, {
    errorProperty: 'errors',
  })

  test('Get plugin installed', async () => {
    await sleep(100)
    await Vue.nextTick()
    await sleep(5)
    expect(document.body).toMatchSnapshot()
  })

  test('Get errors from vue component', async () => {
    const template = {
      data() {
        return {
          t1: false,
          t2: '',
        }
      },
      template: `
      <div>
        <input type="checkbox" name="t1" class="foo" v-model="t1" />
        <input type="radio" name="t2" class="foo" value="foo" v-model="t2"/>
        <input type="radio" name="t2" class="bar" value="bar" v-model="t2"/>
        <span id="age">{{ $errors.first('age') }}</span>
      </div>`,
    }
    const wrapper = mount(template)
    const errors = {
      first_name: ['This field is required'],
      last_name: ['This field is required'],
      age: ['This field is required'],
    }
    wrapper.vm.$errors.fill(errors)
    expect(wrapper.vm.$errors.count()).toEqual(3)
    expect(wrapper.vm.$errors.has('age')).toBeTruthy()
    expect(wrapper.vm.$errors.first('age')).toBe('This field is required')
    expect(template).toMatchSnapshot()
  })
})
