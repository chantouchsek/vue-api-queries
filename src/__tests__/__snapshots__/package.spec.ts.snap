// Vitest Snapshot v1

exports[`Vue Api Queries > Get errors from vue component 1`] = `
{
  "$_vueTestUtils_original": [Circular],
  "_Ctor": {},
  "data": [Function],
  "name": undefined,
  "render": [Function],
  "staticRenderFns": [],
  "template": "
      <div>
        <input type=\\"checkbox\\" name=\\"t1\\" class=\\"foo\\" v-model=\\"t1\\" />
        <input type=\\"radio\\" name=\\"t2\\" class=\\"foo\\" value=\\"foo\\" v-model=\\"t2\\"/>
        <input type=\\"radio\\" name=\\"t2\\" class=\\"bar\\" value=\\"bar\\" v-model=\\"t2\\"/>
        <span id=\\"age\\">{{ $errors.first('age') }}</span>
      </div>",
}
`;

exports[`Vue Api Queries > Get plugin installed 1`] = `
<body>
  <div
    data-app="true"
  />
</body>
`;

exports[`Vue Api Queries Get errors from vue component 1`] = `
Object {
  "$_vueTestUtils_original": [Circular],
  "_Ctor": Object {},
  "data": [Function],
  "name": undefined,
  "render": [Function],
  "staticRenderFns": Array [],
  "template": "
      <div>
        <input type=\\"checkbox\\" name=\\"t1\\" class=\\"foo\\" v-model=\\"t1\\" />
        <input type=\\"radio\\" name=\\"t2\\" class=\\"foo\\" value=\\"foo\\" v-model=\\"t2\\"/>
        <input type=\\"radio\\" name=\\"t2\\" class=\\"bar\\" value=\\"bar\\" v-model=\\"t2\\"/>
        <span id=\\"age\\">{{ $errors.first('age') }}</span>
      </div>",
}
`;

exports[`Vue Api Queries Get plugin installed 1`] = `
<body>
  <div
    data-app="true"
  />
</body>
`;
